### PR TITLE
Dev/options

### PR DESCRIPTION
**Pull request by Kamuso**
In main.js, the option to delete files works, but add/remove tags and set name don't. The request for add/remove tags has a return status of 200 but doesn't seem to do anything, set name gives a 405. I don't know what I'm doing wrong there.
I also need to make the upload page and the fetch tags option before this is finished.

### CODE REVIEW

**Comment by bakape** [comment]
>The request for add/remove tags has a return status of 200

I don't remember testing these. Will dig deeper and see, if it is my fault later this evening. 

>set name gives a 405

Try `/api/images/:id/name`  - without the trailing slash.

**Comment by bakape** [comment]
Yup, tag adding is broken somehow. This will take more time to fix.

**Comment by Kamuso** [comment]
>Try /api/images/:id/name - without the trailing slash.

This made the request work, but the name had a bunch of extra data in it so I've got to fix that. 
>Yup, tag adding is broken somehow. This will take more time to fix.

r.Form.Get("tags") gives a blank string. I'll fix this at the same time as setting the name.

**Comment by Kamuso** [comment]
Okay, everything works as intended now, need to validate input, add fetchtags, and add upload page.

**Comment by bakape** [comment]
> r.Form.Get("tags")

Odd, because that is not the bug I was seeing in the debugger. It went down to db.AddTagsTX() just fine and then something made it not write to the DB. Also, it might be because you are using multipart POST requests there (FormData API). I produced my bug with a PATCH to `http://localhost:8010/api/images/e9336dc8634de6ef577232959e8c526dc7434e50/tags/?tags=sakura_kyouko` style URL. Notice the query string.

**Comment by bakape** [comment]
The tag issue is caused by #54 so just assume it works and I will fix it myself later on.

**Comment by Kamuso** [comment]
I think this will do for now. I wasn't sure how to make dropping images resolve to the import page, so I left that out. Upload progress is shown in the terminal, not the browser, but it seems good enough for now.

File: server.go [hunk 1]
```diff
@@ -298,16 +301,41 @@ func removeTagsHTTP(w http.ResponseWriter, r *http.Request) {
 
 // Set the target file's name
 func setImageNameHTTP(w http.ResponseWriter, r *http.Request) {
-	bytes, err := ioutil.ReadAll(r.Body)
-
+	err := r.ParseForm()
 	if err != nil {
 		sendError(w, 400, err)
```
Comment (bakape) [inline server.go:306]: Missing return on error
```suggestion
		sendError(w, 400, err)
		return
```

File: server.go [hunk 2]
```diff
@@ -298,16 +301,41 @@ func removeTagsHTTP(w http.ResponseWriter, r *http.Request) {
...
+		send500(w, r, err)
+	}
+}
+
+// Fetch tags for a single file
+func fetchTagsHTTP(w http.ResponseWriter, r *http.Request) {
+	sha1 := extractParam(r, "id")
+	pair, err := db.GetImageIDAndMD5(sha1)
+	if err != nil {
+		send500(w, r, err)
```
Comment (bakape) [inline server.go:331]: Same here.
```suggestion
		send500(w, r, err)
		return
```

File: server.go [hunk 3]
```diff
@@ -298,16 +301,41 @@ func removeTagsHTTP(w http.ResponseWriter, r *http.Request) {
...
+	}
+
+	tags, err := fetch.FetchTags(pair.MD5)
+	if err != nil {
+		send500(w, r, err)
```
Comment (bakape) [inline server.go:337]: And here. Yes, we need a better error handling scheme. 
```suggestion
		send500(w, r, err)
		return
```

File: server.go [hunk 4]
```diff
@@ -350,3 +378,37 @@ func serveImagePage(w http.ResponseWriter, r *http.Request) {
...
+func serveImportPage(w http.ResponseWriter, r *http.Request) {
+	setHeaders(w, htmlHeaders)
+	templates.WriteImportPage(w)
+}
+
+// Extract import args and pass to importPaths
+func importPathsHTTP(w http.ResponseWriter, r *http.Request) {
+	err := r.ParseForm()
+	if err != nil {
+		sendError(w, 400, err)
```
Comment (bakape) [inline server.go]:
```suggestion
		send500(w, r, err)
		return
```

File: server.go [hunk 5]
```diff
@@ -350,3 +378,37 @@ func serveImagePage(w http.ResponseWriter, r *http.Request) {
...
+	}
+
+	paths := strings.Split(r.Form.Get("path"), " ")
+	del, err := strconv.ParseBool(r.Form.Get("del"))
+	fetch, err := strconv.ParseBool(r.Form.Get("fetchTags"))
+	if err != nil {
+		sendError(w, 400, err)
```
Comment (bakape) [inline server.go]: Same here.
```suggestion
		send500(w, r, err)
		return
```

**Review by bakape — CHANGES_REQUESTED** [review]

**Comment by bakape** [comment]
Okay. So now we have 2 conflicting ways to upload files in the web UI. Personally I like having a separate upload page like you built more, but automatically navigating to it on image drag&drop would be nice too. Perhaps that can be done by issuing a HTTP redirect to the upload page after all images are uploaded, but not thumbnailed.  But that would also mena we would need to persist uploaded resources cross-request. Using the History API might be better then. Thoughts on this?

>Upload progress is shown in the terminal, not the browser, but it seems good enough for now.

You can use `Transfer-Encoding: chunked` and stream upload progress and any errors to the client on each tick. WIll require calling `.Flush()` on the `http.ResponseWriter` after each tick to prevent buffering. That can be used to display a progress bar. 

Overall - great work. I'll get back to hydron once I'm more or less done with captchouli meguca integration.

**Comment by Kamuso** [comment]
Merry Christmas, have an update.

Mostly fixing some oversights in the import page. 

>2 conflicting ways to upload files in the web UI

I think it would be simplest to remove drag and drop uploads from the browser page and just have it in the imports page. There is a demo of that in this update, so depends how you feel about that.

>You can use `Transfer-Encoding: chunked` and stream upload progress and any errors to the client on each tick. 

I experimented with this but couldn't manage to get it to work. I tried a timed for loop to send some data in batches but it was all sent all together at the end. I don't think I'll be able to get that working on my own.

**Comment by bakape** [comment]
>I think it would be simplest to remove drag and drop uploads from the browser page and just have it in the imports page. There is a demo of that in this update, so depends how you feel about that.

Drag and drop should ideally forward all your dropped files to the upload progress page. The way it is implemented now breaks navigation somewhat.  

>I experimented with this but couldn't manage to get it to work. I tried a timed for loop to send some data in batches but it was all sent all together at the end. I don't think I'll be able to get that working on my own.

Server Sent Events and some JS listening on the frontend? 
That way you can also reuse the current code that displays imports as they are completed. 
https://developer.mozilla.org/en-US/docs/Web/API/Server-sent_events/Using_server-sent_events
https://github.com/alexandrevicenzi/go-sse

**Comment by Kamuso** [comment]
>The way it is implemented now breaks navigation somewhat.

Right, but removing drag&drop from the search page would fix this. Would you accept that instead of redirecting? It would be much simpler.

>Server Sent Events and some JS listening on the frontend?

Turns out SSE isn't necessary, the server code that I had before was correct, the JS wasn't. It now works properly using a `ReadableStream`.

**Comment by bakape** [comment]
> Right, but removing drag&drop from the search page would fix this. Would you accept that instead of redirecting? It would be much simpler.

I can write the redirect part, once we have a complete upload page.

**Comment by Kamuso** [comment]
I'm all done here, then.

**Comment by bakape** [comment]
Okay. I'll take a look at this next weekend.

**Comment by bakape** [comment]
Okay, good work. I merged this into `dev/options`. Once I add my own changes pertaining to drag&drop, I'll merge this into master.

**Comment by Kamuso** [comment]
Added the drag&drop behaviour you wanted. Also did some general clean up, like refactoring the indentation hell in import.js.
Would have pushed to the new pr you made, but I didn't accept the collaboration invite before it expired. tehe pero

**Comment by bakape** [comment]
This has been a very long week.

**Comment by bakape** [comment]
sugoi

**Pull request merged**